Fix URL length calculations for long Permalinks

justinmaurerdotdev opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->
There were a couple of issues present having to do with the calculation of tweet length. 

First, the calculation for the Classic Editor was using the permalink preview to calculate the hypothetical permalink length. That's great until the permalink gets long, and then the preview is shortened with "...". So, the real permalink would be longer than the calculated one. 

Second, Twitter currently shortens all URLs to 23 characters. So, technically, any URL should just be calculated as 23 characters. 

From https://developer.twitter.com/en/docs/counting-characters
> URLs: [All URLs are wrapped in t.co links](https://developer.twitter.com/content/developer-twitter/en/docs/basics/tco). This means a URL’s length is defined by the transformedURLLength parameter in the [twitter-text configuration file](https://github.com/twitter/twitter-text/tree/master/config). The current length of a URL in a Tweet is 23 characters, even if the length of the URL would normally be shorter.

From https://developer.twitter.com/en/docs/tco, evidently there's an API endpoint that returns the current maximum URL length. In some cases, shorter URLs may not reach the maximum length, but in all cases, the maximum is the maximum. 

In this PR, I: 
- Set up a new constant, `AUTOSHARE_FOR_TWITTER_URL_LENGTH`, to store the current URL length published by Twitter. This could be set to false to revert to using the real URL lengths. This could be refactored to get the most current value from the `help/configuration` endpoint, as [described here](https://developer.twitter.com/en/docs/tco).
- Localized that Twitter URL length value in `localize_data()` as `adminAutoshareForTwitter.twitterURLLength`.
- Added a way of calculating the accurate hypothetical permalink length for the Classic Editor (I haven't worked with Gutenberg much, and did not spend any time checking the implementation there).
- Set the default behavior everywhere to use the twitterURLLength value _instead_ of the actual permalink length. twitterURLLength can be set to false in the `localize_data()` implementation to force real permalink calculations (because each implementation falls back to real permalink calculation when the `twitterURLLength` is false.
- Modified the `compose_tweet_body()` method to also use the defined twitter URL length for its calculations.

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->
To be honest, I am not familiar with the Cypress framework enough to write an automated test. And I'm not sure if all of the files I've touched are in scope for those types of tests anyway. I would be happy to help pair-program tests for this, or dialogue about them here.

To manually test the correct functionality, there are two scenarios to test: 1) the correct default, which is to treat all URLs as the same length and 2) the fallback, in case we want to disable that feature. 

**To test the fallback**
- [ ] Set the `AUTOSHARE_FOR_TWITTER_URL_LENGTH` value to `false` in the main plugin file. 
- [ ] Start a new post. 
- [ ] Write a title that exceeds WordPress's permalink preview length (e.g. "This is my super duper, really long post title that definitely isn't going to fit in the preview"). 
- [ ] The calculated tweet length should be exactly the length of the intended permalink plus the 5-character padding. 
- [ ] Then try to write a custom tweet. 
- [ ] The tweet body validation should properly use the real permalink length, so that the calculated length is the true permalink + the length of the message + the 5-character padding.

**To test the currently-intended behavior**
- [ ] Set the `AUTOSHARE_FOR_TWITTER_URL_LENGTH` value back to `23` in the main plugin file. 
- [ ] Start a new post. 
- [ ] Write a title that exceeds WordPress's permalink preview length (e.g. "This is my super duper, really long post title that definitely isn't going to fit in the preview"). 
- [ ] Try to customize the tweet with content that is exactly 252 characters long. 280 - 23 (the current maximum URL length for tweets) - 5 (the pre-existing padding used for the plugin). 
- [ ] The tweet should be successful/valid at 252 characters, no matter what the URL length.

This should work in Classic Editor and Gutenberg (I think, but wasn't sure about the plugin build process). In both cases, the Tweet should successfully send without errors when the post is published.

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Fixed - Bug fix for issues calculating tweet length


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @justinmaurerdotdev


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [see my comments above] I have added tests to cover my change.
- [?] All new and existing tests pass. 
I get `Error: ENOENT: no such file or directory, scandir '/Users/justinmaurer/.wp-env'` When I try to run the `cypress:run` script. I'm not really familiar with `.wp-env` and don't know what's expected to be in there (or how to set it up).

I'm aware there is probably some discussion or cleanup needed to make my commits fit the package standards, but I wanted to submit it at this stage to get some feedback about what exactly is missing, and whether there's some other plan (or better way) to handle this.

justinmaurerdotdev committed:
Added a function for the classic editor that properly calculates the permalink length. Define the "transformed" URL length, according to the Twitter docs: https://developer.twitter.com/en/docs/counting-characters. Added "twitterURLLength" to the adminAutoshareForTwitter localized object. Refactored classic editor and block editor codebases to prefer the "twitterURLLength" value, if available.

iamdharmesh commented:
@justinmaurerdotdev Thank you for the PR and for sharing useful information on how Twitter counts URL characters. I will get this reviewed very soon.

justinmaurerdotdev committed:
phpcs style fix

iamdharmesh commented on assets/js/admin-autoshare-for-twitter-classic-editor.js:
```diff
@@ -125,15 +125,41 @@
+				// according to this page https://developer.twitter.com/en/docs/counting-characters, all URLs are transformed to a uniform length
+				permalinkLength = adminAutoshareForTwitter.twitterURLLength;
+			} else {
+				// The #sample-permalink > a tag seems like the only convenient place to get the correct path for the post,
+				// without doing an AJAX call. We're using this technique to isolate the path (remove the <span>):
+				// https://stackoverflow.com/a/74517817
+				let slug = $("#editable-post-name-full").text();
+
+				// get the hypothetical path
+				let aTagContents = $("#sample-permalink > a")[0].innerHTML;
+				let workingDiv = document.createElement("span");
+				workingDiv.innerHTML = aTagContents;
+				let fakeSlugSpan = workingDiv.querySelector('span');
+				workingDiv.removeChild(fakeSlugSpan);
+				let permalinkPrefix = workingDiv.innerText;
+				let permalink = permalinkPrefix + slug + '/';
+				permalinkLength = permalink.length;
```
Just curious, if we can use `href` attribute of #sample-permalink > a tag here to get permalink length.
```js
permalinkLength = jQuery('#sample-permalink a').attr("href").length
```

justinmaurerdotdev replied:
Unfortunately, this link is populated with the preview link until the post is published, so something like "https://mysite.test/?p=8129122&preview=true" will be there instead of the actual URL.

iamdharmesh commented on assets/js/admin-autoshare-for-twitter-classic-editor.js:
```diff
@@ -125,15 +125,41 @@
 		let permalinkLength = 0;
 		if ( $('#sample-permalink').length ) {
-			permalinkLength = $('#sample-permalink').text().length
+			if (adminAutoshareForTwitter.twitterURLLength) {
```
As all URLs are wrapped in t.co links. I think we can remove the fallback logic here. However, I noticed that Twitter doesn't wrap inaccessible URLs (local wp setup), for those cases we may still want to go with the current length counter logic.

So, the best way here is to keep our current url chars counter for local setups and Fix 23 chars in live sites. what do you think?

iamdharmesh replied:
We can add a function given below in Utils.php to identify the wp environment type.
```php
/**
 * Returns whether WP environment is local.
 *
 * @return boolean
 */
function is_local() {
	$is_local_env = in_array( wp_get_environment_type(), [ 'local', 'development' ], true );
	$is_local_url = strpos( home_url(), '.test' ) || strpos( home_url(), '.local' );
	$is_local     = $is_local_env || $is_local_url;

	/**
	 * Filters whether WP environment is local or not.
	 *
	 * @param bool   whether WP environment is local or not.
	 */
	return apply_filters( 'autoshare_for_twitter_is_local', $is_local );
}

```

justinmaurerdotdev replied:
Ah, interesting. Yes, that does make sense.

iamdharmesh reviewed (changes requested):
@justinmaurerdotdev Thank you for the PR and very detailed information. This is amazing. We may need to change a bit on how we want to keep using the current URL length counter for local setups instead of identifying using fallback based on the value of `AUTOSHARE_FOR_TWITTER_URL_LENGTH` constant. Otherwise, All looks good to me. I have added some comments on the code to check.

Thank you.

justinmaurerdotdev committed:
- disable URL shortening calculations when using a test environment, because Twitter doesn't shorten inaccessible links

justinmaurerdotdev commented:
I've refactored this a bit to have the `AUTOSHARE_FOR_TWITTER_URL_LENGTH` constant itself set as the plugin loads, so that that constant can be the source of truth for whether the shortening should happen, _and_ what length to use (integer for length, otherwise it should be `false`). The switch is dependent on the new `is_local()` function, added to utils.php, as suggested.

justinmaurerdotdev committed:
force JS global URL length value to be interpreted as a Number

justinmaurerdotdev commented:
Sorry for the extra notifications today. Was able to test in both types of environments and realized the JS was sometimes doing string concatenation instead of numeric addition. Otherwise, I'm pretty sure this is (at least functionality-wise) ready to go. I'm not sure what it means that I'm assigned on this PR, but let me know if you need anything from me.

iamdharmesh committed:
Minor tweak on usage of fixed url length and fixed additional slash(/) issue in classic editor length calculation.

iamdharmesh reviewed (approved):
Hi @justinmaurerdotdev, Sorry for the delay in the review. Thanks for making the additional changes. PR looks amazing now. I have made minor tweaks to the usage of fixed URL length and additional slash(/) in URL length calculation in the classic editor. PR is ready to merge. Please let me know if you see anything wrong with the tweaks I made, I would love to fix it.

Thanks again for this great contribution.

github-actions[bot] committed:
Merge branch 'develop' into fix-url-length-calculations

justinmaurerdotdev commented:
Was out of the office last week, but this looks good to me. Very glad to see this fixed. Thanks for working with me on it!
